spark/src/main/scala/geotrellis/spark/io/FilteringLayerReader.scala

```diff
@@ -1,17 +1,26 @@
 package geotrellis.spark.io
 
 import geotrellis.spark.Boundable
+import geotrellis.spark.io.index.KeyIndex
+import geotrellis.spark.io.json._
+import spray.json.JsonFormat
 
-abstract class FilteringLayerReader[ID, K: Boundable, M, ReturnType] extends LayerReader[ID, ReturnType] {
+abstract class FilteringLayerReader[ID, K: Boundable, M, ReturnType] extends LayerReader[ID, K, ReturnType] {
 
-  def read(id: ID, rasterQuery: RDDQuery[K, M], numPartitions: Int): ReturnType
+  def read[I <: KeyIndex[K]: JsonFormat](id: ID, rasterQuery: RDDQuery[K, M], numPartitions: Int, format: JsonFormat[I]): ReturnType
+
+  def read[I <: KeyIndex[K]: JsonFormat](id: ID, numPartitions: Int, format: JsonFormat[I]): ReturnType =
```

Comment:
I very much like the idea of `SomethingLayerReader(path).withKeyIndexFormat(customKeyIndexFormatter)`, fluent APIs are hot right now.

I would however keep the machinery of extending the KeyIndexFormat as simple as possible. Having `KeyIndexFormat.addType` do an implicit resolution doesn't seems necessary. At this point the user is extending a known thing with a known thing at a specific point before passing in a concrete object to the reader.